superset/assets/src/visualizations/core/models/Preset.js

```diff
@@ -0,0 +1,29 @@
+import Plugin from './Plugin';
+
+export default class Preset {
+  constructor({
+    name = '',
+    description = '',
+    presets = [],
+    plugins = [],
+  } = {}) {
+    this.name = name;
+    this.description = description;
+    this.presets = presets;
+    this.plugins = plugins;
+  }
+
+  install() {
+    this.presets.forEach((preset) => {
+      preset.install();
+    });
+    this.plugins.forEach((plugin) => {
+      if (plugin instanceof Plugin) {
+        plugin.install();
+      } else {
+        plugin();
```

Comment:
Could you explain a bit when plugin will be just a function ?

Reply:
could you address this by having the `Plugin` take the arguments via the constructor instead?

Reply:
Aha. good idea.

Reply:
Umm, actually can't, because each vis will declare the ChartPlugin instance and export, so the preset only have access to the instance.

barPlugin.js
```js
const plugin = new ChartPlugin(...);
export default plugin;
```

Reply:
Umm, on second thought, I can add `.setConfig` to the `Plugin` class

```
plugins: [
  plugin1,
  plugin2.setConfig({ key: 'abc' }),
]
```

Reply:
hmm, could the pattern instead be to export a class that has extended `ChartPlugin`?

```js
// BarPlugin.js
class BarPlugin extends ChartPlugin {
  ...
}

export default BarPlugin;

// preset.js
import BarPlugin from './BarPlugin';
...
const plugins = [
  new BarPlugin({ config: 123 }),
];
...
```

Reply:
That could work too. At first I was like hmm is creating new class an overkill. Perhaps not. Just a bit more typing. and that allows override `.register()` function + more customization.

```js
class BarPlugin extends ChartPlugin {
  constructor(config) {
    super({ config, metadata, Chart, transformProps });
  }
}
```

vs

```js
const barPlugin = new ChartPlugin({ metadata, Chart, transformProps });
barPlugin.configure(config);
```